R/busy-indicators.R

```diff
@@ -0,0 +1,186 @@
+#'   be any valid CSS time and can useful for not showing the spinner
+#'   if the computation finishes quickly.
+#' @param css_selector A CSS selector for scoping the spinner customization.
+#'   Defaults to the root element.
+#'
+#' @export
+#' @seealso [useBusyIndicators()] [pulseOptions()]
+#' @examplesIf interactive()
+#'
+#' library(bslib)
+#'
+#' ui <- page_fillable(
+#'   useBusyIndicators(),
+#'   spinnerOptions(color = "orange"),
+#'   card(
+#'     card_header(
+#'       "A plot",
+#'       input_task_button("simulate", "Simulate"),
+#'       class = "d-flex justify-content-between align-items-center"
+#'     ),
+#'     plotOutput("p"),
+#'   )
+#' )
+#'
+#' server <- function(input, output) {
+#'   output$p <- renderPlot({
+#'     input$simulate
+#'     Sys.sleep(4)
+#'     plot(x = rnorm(100), y = rnorm(100))
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+spinnerOptions <- function(
+  type = NULL,
+  ...,
+  color = NULL,
+  size = NULL,
+  easing = NULL,
+  speed = NULL,
+  delay = NULL,
+  css_selector = ":root"
+) {
+
+  # bounce requires a different animation than the others
+  if (isTRUE(type == "bounce")) {
+    animation <- "shiny-busy-spinner-bounce"
+    speed <- speed %||% "0.8s"
+  } else {
+    animation <- NULL
+  }
+
+  # Supported types have a CSS var already defined with their SVG data
+  if (isTRUE(type %in% c("default", "tadpole", "disc", "dots", "dot-track", "bounce"))) {
+    type <- sprintf("var(--_shiny-spinner-type-%s)", type)
+  }
+
+  # Options are controlled via CSS variables.
+  css_vars <- paste0(c(
+    if (!is.null(type)) sprintf("--shiny-spinner-mask-img: %s", type),
+    if (!is.null(easing)) sprintf("--shiny-spinner-easing: %s", easing),
+    if (!is.null(animation)) sprintf("--shiny-spinner-animation: %s", animation),
+    if (!is.null(color)) sprintf("--shiny-spinner-color: %s", color),
+    if (!is.null(size)) sprintf("--shiny-spinner-size: %s", size),
+    if (!is.null(speed)) sprintf("--shiny-spinner-speed: %s", speed),
+    if (!is.null(delay)) sprintf("--shiny-spinner-delay: %s", delay)
+  ), collapse = ";")
+
+  # The CSS cascade allows this to be called multiple times, and as long as the CSS
+  # selector is the same, the last call takes precedence. Also, css_selector allows
+  # for scoping of the spinner customization.
+  tags$style(HTML(paste0(css_selector, " {", css_vars, "}")))
+}
+
+
+#' Customize the pulsing busy indicator.
+#'
+#' Include the result of this function in the app's UI to customize the pulsing
+#' banner.
+#'
+#' @param color The color of the pulsing banner. This can be any valid CSS
+#'   color. Defaults to the app's "primary" color (if Bootstrap is on the page)
+#'   or light-blue if not.
+#' @param height The height of the pulsing banner. This can be any valid CSS
+#'   size. Defaults to "3.5px".
+#' @export
+#' @seealso [useBusyIndicators()] [spinnerOptions()]
+pulseOptions <- function(color = NULL, height = NULL, speed = NULL) {
```

Comment:
Add `css_selector`

Reply:
Actually we don't need this since there's only one pulse (on the root of the doc)